src/mscorlib/src/System/Threading/ExecutionContext.cs

```diff
@@ -191,18 +191,27 @@ internal static void SetLocalValue(IAsyncLocal local, object newValue, bool need
             if (previousValue == newValue)
                 return;
 
-            current = new ExecutionContext(current);
-            current.m_localValues[local] = newValue;
-
-            t_currentMaybeNull = current;
+            Dictionary<IAsyncLocal, object> newValues = new Dictionary<IAsyncLocal, object>(current.m_localValues);
+            newValues[local] = newValue;
 
+            List<IAsyncLocal> newChangeNotifications = current.m_localChangeNotifications;
             if (needChangeNotifications)
             {
                 if (hadPreviousValue)
-                    Contract.Assert(current.m_localChangeNotifications.Contains(local));
+                {
+                    Contract.Assert(newChangeNotifications.Contains(local));
+                }
                 else
-                    current.m_localChangeNotifications.Add(local);
+                {
+                    newChangeNotifications = new List<IAsyncLocal>(current.m_localChangeNotifications);
+                    newChangeNotifications.Add(local);
+                }
+            }
+
+            t_currentMaybeNull = new ExecutionContext(newValues, newChangeNotifications);
```

Comment:
But the new ExecutionContext would be uncaptured? The scenario would be if you set more than 1 AsyncLocal value; each set would create a new context, however the old one isn't kept anywhere so is immediately garbage; whereas it could be reused for each set; until its been captured and thus frozen?